runtime/Java/src/org/antlr/v4/runtime/atn/PredictionContext.java

```diff
@@ -67,19 +63,19 @@ protected PredictionContext(int cachedHashCode) {
 	}
 
 	/** Convert a {@link RuleContext} tree to a {@link PredictionContext} graph.
-	 *  Return {@link #EMPTY} if {@code outerContext} is empty or null.
+	 *  Return {@link EmptyPredictionContext#Instance} if {@code outerContext} is empty or null.
 	 */
 	public static PredictionContext fromRuleContext(ATN atn, RuleContext outerContext) {
-		if ( outerContext==null ) outerContext = RuleContext.EMPTY;
+		if ( outerContext==null ) outerContext = ParserRuleContext.EMPTY;
 
 		// if we are in RuleContext of start rule, s, then PredictionContext
 		// is EMPTY. Nobody called us. (if we are empty, return empty)
-		if ( outerContext.parent==null || outerContext==RuleContext.EMPTY ) {
-			return PredictionContext.EMPTY;
+		if ( outerContext.parent==null || outerContext==ParserRuleContext.EMPTY ) {
```

Comment:
same here. maybe shouldn't be ParserRuleContext?

Reply:
Possible supertypes of: `RuleContext`:

![image](https://user-images.githubusercontent.com/1150330/174674020-9d27bee0-10c9-450a-a7c2-10008ef31cc5.png)

Moreover, all concrete classes are supertypes of `ParseRuleContext`, not `RuleContext`. I don't understand meaning of `RuleContext`, it looks redundant.

Reply:
I think lexer uses that. Parser tracks things to build parse tree.